File: 5_longest_palindromic_substring/main.cpp
/*
 * Given a string s, find the longest palindromic substring in s.
 * You may assume that the maximum length of s is 1000.

 * Example 1:
 * Input: "babad"
 * Output: "bab"
 * Note: "aba" is also a valid answer.
 *
 * Example 2:
 * Input: "cbbd"
 * Output: "bb"
 */

#include <iostream>
// #include <string>

class Solution {
public:
    std::string longestPalindrome(std::string s) {
	if (s.size() < 2) return s;

	int start = 0, max_len = 1, new_len = 0;
	int left, right;
	int i = 0;
	while (i < s.size()) {
	    if (s.size() - i <= max_len / 2) break;
	    
	    left = i;
	    right = i;
	    while (right < s.size() - 1 && s[right + 1] == s[right]) ++right;
	    i = right + 1;

	    while (right < s.size() - 1 && left > 0 && s[right + 1] == s[left - 1]) {
		++right;
		--left;
	    }

	    new_len = right - left + 1;
	    if (new_len > max_len) {
		start = left;
		max_len = new_len;
	    }
	}

	return s.substr(start, max_len);
    }
};

int main() {
    Solution sl;

    std::string s = "babad";
    std::cout << sl.longestPalindrome(s) << std::endl;

    s = "cbbd";
    std::cout << sl.longestPalindrome(s) << std::endl;
}

File: 5_longest_palindromic_substring/go/main.go
package main

import (
	"fmt"
)

func longestPalindrome(s string) string {
	if len(s) < 2 {
		return s
	}

	var (
		left    int
		right   int
		new_len int
		start   int
	)
	i, max_len := 0, 1

	for i < len(s) {
		if len(s) - i <= max_len / 2 {
			break
		}

		left, right = i, i
		for right < len(s) - 1 && s[right] == s[right + 1] {
			right++
		}
		i = right + 1

		for left > 0 && right < len(s) - 1 && s[left - 1] == s[right + 1] {
			left--
			right++
		}

		new_len = right - left + 1
		if new_len > max_len {
			start = left
			max_len = new_len
		}
	}
	return s[start:start + max_len]
}

func main() {
	s := "babad"
	fmt.Println(longestPalindrome(s))

	s = "cbbd"
	fmt.Println(longestPalindrome(s))
}
